src/core/symbology/classification/qgsclassificationmethod.cpp

```diff
@@ -0,0 +1,246 @@
+
+  const QString lowerLabel = valueToLabel( lowerValue );
+  const QString upperLabel = valueToLabel( upperValue );
+
+  QString label( mLabelFormat );
+  label.replace( QLatin1String( "%1" ), lowerLabel ).replace( QLatin1String( "%2" ), upperLabel );
+  return label;
+}
+
+QgsClassificationMethod *QgsClassificationMethod::createFromXml( const QDomElement &elem )
+{
+  const QString methodId = elem.attribute( QStringLiteral( "id" ) );
+  QgsClassificationMethod *method = QgsApplication::classificationMethodRegistry()->method( methodId );
+
+  return method;
+}
+
+QDomElement QgsClassificationMethod::writeExtraXml( QDomDocument &doc ) const
+{
+  return QDomElement();
+}
+
+void QgsClassificationMethod::setSymmetricMode( bool enabled, double symmetryPoint, bool astride )
+{
+  mSymmetricEnabled = enabled;
+  mSymmetryPoint = symmetryPoint;
+  mAstride = astride;
+}
+
+void QgsClassificationMethod::setLabelPrecision( int precision )
+{
+  // Limit the range of decimal places to a reasonable range
+  precision = qBound( MIN_PRECISION, precision, MAX_PRECISION );
+  mLabelPrecision = precision;
+  mLabelNumberScale = 1.0;
+  mLabelNumberSuffix.clear();
+  while ( precision < 0 )
+  {
+    precision++;
+    mLabelNumberScale /= 10.0;
+    mLabelNumberSuffix.append( '0' );
+  }
+}
+
+QString QgsClassificationMethod::formatNumber( double value ) const
+{
+  if ( mLabelPrecision > 0 )
+  {
+    QString valueStr = QLocale().toString( value, 'f', mLabelPrecision );
+    if ( mLabelTrimTrailingZeroes )
+      valueStr = valueStr.remove( RE_TRAILING_ZEROES );
+    if ( RE_NEGATIVE_ZERO.exactMatch( valueStr ) )
+      valueStr = valueStr.mid( 1 );
+    return valueStr;
+  }
+  else
+  {
+    QString valueStr = QLocale().toString( value * mLabelNumberScale, 'f', 0 );
+    if ( valueStr == QLatin1String( "-0" ) )
+      valueStr = '0';
+    if ( valueStr != QLatin1String( "0" ) )
+      valueStr = valueStr + mLabelNumberSuffix;
+    return valueStr;
+  }
+}
+
+QList<QgsClassificationRange> QgsClassificationMethod::classes( const QgsVectorLayer *vl, const QString &expression, int numberOfClasses )
+{
+  if ( expression.isEmpty() )
+    return QList<QgsClassificationRange>();
+
+  if ( numberOfClasses < 1 )
+    numberOfClasses = 1;
+
+  QList<double> values;
+  double minimum;
+  double maximum;
+
+  int fieldIndex = vl->fields().indexFromName( expression );
+
+  bool ok;
+  if ( mValuesRequired || fieldIndex == -1 )
+  {
+    values = QgsVectorLayerUtils::getDoubleValues( vl, expression, ok );
+    if ( !ok || values.isEmpty() )
+      return QList<QgsClassificationRange>();
+
+    auto result = std::minmax_element( values.begin(), values.end() );
+    minimum = *result.first;
+    maximum = *result.second;
+  }
+  else
+  {
+    minimum = vl->minimumValue( fieldIndex ).toDouble();
+    maximum = vl->maximumValue( fieldIndex ).toDouble();
+  }
+
+  // get the breaks
+  const QList<double> breaks = calculateBreaks( minimum, maximum, values, numberOfClasses );
```

Comment:
Just had a quick glance over -- one thing that strikes me immediately is that there's potentially a huge regression in speed for some methods here. E.g. the existing ones like equal interval, which only require a min and max value from a layer will now be significantly slower as they require a full layer iteration.

Could this be avoided, e.g. via some flags the subclass could return to say "i only need min and max" or "i need all values"?

Reply:
yep, already implemented ;)
Look at the non-virtual public method in QgsClassificationMethod.
And every implementation can define if values are required when calling the parent constructor.

Reply:
but I kept only one pure virtual private method to be implemented to make things simpler.